src/mapping/GreedyMapping.hpp

```diff
@@ -0,0 +1,293 @@
+  virtual void mapConservative(const time::Sample &inData, Eigen::VectorXd &outData) = 0;
+  virtual void mapConsistent(const time::Sample &inData, Eigen::VectorXd &outData) = 0;
+  virtual void clear() override;
+
+protected:
+  using RadialBasisFctBaseMapping<RADIAL_BASIS_FUNCTION_T>::_basisFunction;
+
+  bool _usesPolynomial;
+
+  mesh::PtrMesh _inputMesh;
+  mesh::PtrMesh _outputMesh;
+
+  size_t _maxIter;
+  double _tolerance;
+
+  size_t _inSize  = 0;
+  size_t _outSize = 0;
+  size_t _basisSize = 0;
+
+  std::vector<int>    _greedyIDs;
+  std::array<bool, 3> _activeAxis;
+
+  Eigen::MatrixXd _kernelEval;
+  Eigen::MatrixXd _polyMatrixQ;
+  Eigen::MatrixXd _polyMatrixU;
+
+  Eigen::ColPivHouseholderQR<Eigen::MatrixXd> _qrDecomposedQ;
+
+  std::pair<int, double> select(const Eigen::VectorXd &powerFunction) const;
+  std::pair<int, double> select(const Eigen::MatrixXd &residual) const;
+
+  void fillEvaluationMatrix();
+  void fillPolynomialMatrices();
+
+  template <typename IndexContainer>
+  void updateKernelVector(const mesh::Vertex &x, const IndexContainer &ids, Eigen::VectorXd &kernelVector) const;
+
+  void solveConservativeWithCut(const time::Sample &inData, const Eigen::MatrixXd &cut, Eigen::VectorXd &outData) const;
+  void solveConsistentWithCut(const time::Sample &inData, const Eigen::MatrixXd &cut, Eigen::VectorXd &outData) const;
+  void solveConservativeWithCholesky(const time::Sample &inData, const Eigen::MatrixXd &choleskyA, Eigen::VectorXd &outData) const;
+  void solveConsistentWithCholesky(const time::Sample &inData, const Eigen::MatrixXd &choleskyA, Eigen::VectorXd &outData) const;
+
+  size_t estimateNumberOfCenters();
+};
+
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+GreedyMapping<RADIAL_BASIS_FUNCTION_T>::GreedyMapping(
+    Mapping::Constraint     constraint,
+    int                     dimensions,
+    RADIAL_BASIS_FUNCTION_T function,
+    std::array<bool, 3>     deadAxis,
+    Polynomial              polynomial,
+    GreedyParameter         greedyParameter)
+    : RadialBasisFctBaseMapping<RADIAL_BASIS_FUNCTION_T>(constraint, dimensions, function, deadAxis, Mapping::InitialGuessRequirement::None)
+{
+  PRECICE_ASSERT(polynomial != Polynomial::ON, "Integrated polynomials not supported for greedy rbf methods");
+  _usesPolynomial = (polynomial == Polynomial::SEPARATE);
+
+  _tolerance = greedyParameter.tolerance;
+  _maxIter   = greedyParameter.maxIterations;
+
+  _activeAxis = std::array<bool, 3>({{false, false, false}});
+  std::transform(deadAxis.begin(), deadAxis.end(), _activeAxis.begin(), [](const auto ax) { return !ax; });
+}
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+void GreedyMapping<RADIAL_BASIS_FUNCTION_T>::fillPolynomialMatrices() {
+
+  precice::profiling::Event e("fillPolynomialMatrices", profiling::Synchronize);
+
+  unsigned int polyParams = 4 - std::count(_activeAxis.begin(), _activeAxis.end(), false);
+  _polyMatrixQ.resize(_greedyIDs.size(), polyParams);
```

Comment:
```suggestion
  _polyMatrixQ.resize(_greedyIDs.size(), polyParams);
```
Why do we use here the expansion size? I would consider the polynomial interpolation a pre-processing step and we can easily use all input mesh data points. The QR-decomposition is anyway cheap.

Reply:
Good question. I think it was more convenient to implement. I will change that.